src/clojure/nrepl/middleware/dynamic_loader.clj

```diff
@@ -0,0 +1,108 @@
+                                           :status :done}))
+      (h msg))))
+
+(set-descriptor! #'wrap-ping
+                 {:require #{}
+                  :expects #{}
+                  :handles {"ping"
+                            {:doc "Ping"
+                             :require {}
+                             :returns {"status" "done"}}}})
+
+(defn unknown-op
+  "Sends an :unknown-op :error for the given message."
+  [{:keys [op transport] :as msg}]
+  (t/send transport (response-for msg :status #{:error :unknown-op :done} :op op)))
+
+(defn- update-stack!
+  [session middlewares]
+  (let [ctxcl  (.getContextClassLoader (Thread/currentThread))
+        alt-cl (when-let [classloader (:classloader (meta session))]
+                 (classloader))
+        cl     (or alt-cl
+                   ctxcl)]
+    (.setContextClassLoader (Thread/currentThread) cl)
+    (try
+      (with-bindings {clojure.lang.Compiler/LOADER cl}
+        (let [stack (->> middlewares
+                         (map (fn [middleware]
+                                (require-resolve
+                                 (if (var? middleware)
+                                   (str/replace (str middleware) "#'" "")
+                                   (str/replace middleware "#'" "")))))
+                         middleware/linearize-middleware-stack)]
+          (reset! *state* {:handler ((apply comp (reverse stack)) unknown-op)
+                           :stack   stack})))
+
+      (catch Throwable t
+        {:error        t
+         :alt-cl?      (boolean alt-cl)
+         :before-stack (map str (:stack @*state*))})
+      (finally
+        (.setContextClassLoader (Thread/currentThread) ctxcl)))))
+
+(defn wrap-dynamic-loader
+  [h]
+  (fn [{:keys [op transport session middlewares] :as msg}]
+    (case op
+      "add-middlewares"
+      (do
+        (let [{:keys [error alt-cl? before-stack]}
+              (update-stack! session (concat middlewares (:stack @*state*)))]
+          (when error
+            (t/send transport (response-for msg {:error        (str error)
+                                                 :sideloading  (str alt-cl?)
+                                                 :before-stack before-stack}))))
+        (t/send transport (response-for msg {:status :done})))
+
+      "swap-middlewares"
+      (do
+        (update-stack! session middlewares)
+        (when transport
+          (t/send transport (response-for msg :middlewares (mapv str (:stack @*state*))))
+          (t/send transport (response-for msg {:status :done}))))
+
+      "ls-middlewares"
+      (do
+        (t/send transport (response-for msg :middlewares (mapv str (:stack @*state*))))
+        (t/send transport (response-for msg {:status :done})))
+
+      (h msg))))
+
+(set-descriptor! #'wrap-dynamic-loader
+                 {:requires #{#'middleware.session/session}
+                  :expects  #{}
+                  :handles  {"add-middlewares"
```

Comment:
Seems you forgot to list the other ops here.

Reply:
Yes. Will add these in when I do the documentation